File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/filter/RunMoveMustBeFollowedBySomethingUsefulFilter.java
package org.silnith.game.solitaire.move.filter;

import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import org.silnith.deck.Card;
import org.silnith.game.GameState;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.move.SolitaireMove;

/**
 * If a run is moved from one column to another, one of two conditions must hold true.
 * Either the entire run is moved, exposing a new face-down card (or emptying a column),
 * or the subsequent move must involve the existing face-up card exposed by moving the run.
 */
public class RunMoveMustBeFollowedBySomethingUsefulFilter implements SolitaireMoveFilter {

    @Override
    public Object getStatisticsKey() {
        return "Run Move Must Be Useful";
    }

	@Override
	public boolean shouldFilter(final List<GameState<SolitaireMove, Board>> gameStateHistory) {
	    final Iterator<GameState<SolitaireMove, Board>> iterator = gameStateHistory.iterator();
	    assert iterator.hasNext();
	    final GameState<SolitaireMove, Board> currentGameState = iterator.next();
		final SolitaireMove currentMove = currentGameState.getMove();
		assert currentMove != null;
		
		if (!iterator.hasNext()) {
            /*
             * This can only happen at the very beginning of the game.
             * In that case, this filter is not helpful, so just let everything pass.
             */
		    return false;
		}
		
		final GameState<SolitaireMove, Board> previousGameState = iterator.next();
		final SolitaireMove previousMove = previousGameState.getMove();
		final Board previousBoard = previousGameState.getBoard();
		/*
		 * This is the board AFTER the move has been applied!
		 */
		
		if (previousMove.isFromColumn()) {
            assert previousMove.hasCards();
			// check whether the previous run move used up all the cards
			// if it did, everything is fine

			if (!iterator.hasNext()) {
			    /*
			     * This can only happen at the very beginning of the game.
			     * In that case, this filter is not helpful, so just let everything pass.
			     */
			    return false;
			}
			final GameState<SolitaireMove, Board> gameStateTwoStepsBack = iterator.next();
			final Board boardTwoStepsBack = gameStateTwoStepsBack.getBoard();
			
            /*
             * Check whether the previous move took all the available cards from the source column.
             */
            final int sourceColumnIndex = previousMove.getFromColumnIndex();
            final int numberOfMovedCards = previousMove.getCards().size();
			if (numberOfMovedCards == boardTwoStepsBack.getColumn(sourceColumnIndex).getNumberOfFaceUpCards()) {
			    /*
			     * The previous move took all face-up cards from the source column.
			     * Therefore, it is not suspicious.
			     * (Shenanigans with the king are handled by other filters.)
			     */
				return false;
			} else {
			    /*
			     * The previous move took only a portion of the available run.
			     * Therefore, we want to make sure the top-most card NOT taken
			     * is involved with the current move.  Otherwise, the previous
			     * move has no value.
			     */
				final Card cardExposedByRunMove = previousBoard.getColumn(sourceColumnIndex).getTopCard();
				if (Collections.singletonList(cardExposedByRunMove).equals(currentMove.getCards())) {
					// This move uses the exposed card, meaning the previous run move is allowed.
					// This makes sure the move uses ONLY the exposed card, not as one of a run.
					// TODO: Should the only allowed follow-up move be column to foundation?
					return false;
				} else {
					/*
					 * This move does not use the exposed card, so filter the sequence of moves.
					 */
					return true;
				}
			}
		} else {
			/*
			 * Not interested in any scenario except where the second-most-recent move is moving a run from one column to another.
			 */
			return false;
		}
	}

}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/move/filter/RunMoveMustBeFollowedBySomethingUsefulFilter.cpp
#include <silnith/game/solitaire/move/filter/RunMoveMustBeFollowedBySomethingUsefulFilter.h>

using namespace std;
using namespace std::literals::string_literals;

namespace silnith::game::solitaire::move::filter
{
    string RunMoveMustBeFollowedBySomethingUsefulFilter::get_statistics_key() const
    {
        return "Run Move Must Be Useful"s;
    }

    bool RunMoveMustBeFollowedBySomethingUsefulFilter::should_filter(std::shared_ptr<linked_node<game_state<solitaire_move, board>>> const& game_state_history) const
    {
        linked_node<game_state<solitaire_move, board>>::const_iterator iterator{ game_state_history->cbegin() };
        linked_node<game_state<solitaire_move, board>>::const_iterator end{ game_state_history->cend() };
        game_state<solitaire_move, board> current_game_state{ *iterator };
        iterator++;
        shared_ptr<solitaire_move> current_move{ current_game_state.get_move() };
        shared_ptr<board> current_board{ current_game_state.get_board() };

        if (iterator == end)
        {
            /*
             * This can only happen at the very beginning of the game.
             * In that case, this filter is not helpful, so just let everything pass.
             */
            return false;
        }

        game_state<solitaire_move, board> previous_game_state{ *iterator };
        iterator++;
        shared_ptr<solitaire_move> previous_move{ previous_game_state.get_move() };
        shared_ptr<board> previous_board{ previous_game_state.get_board() };
        /*
         * This is the board AFTER the move has been applied!
         */

        if (previous_move->is_from_column())
        {
            // check whether the previous run move used up all the cards
            // if it did, everything is fine

            if (iterator == end)
            {
                /*
                 * This can only happen at the very beginning of the game.
                 * In that case, this filter is not helpful, so just let everything pass.
                 */
                return false;
            }
            game_state<solitaire_move, board> game_state_two_steps_back{ *iterator };
            iterator++;
            shared_ptr<board> board_two_steps_back{ game_state_two_steps_back.get_board() };

            /*
             * Check whether the previous move took all the available cards from the source column.
             */
            size_t source_column_index{ previous_move->get_source_column_index() };
            size_t number_of_moved_cards{ previous_move->get_cards().size() };
            if (number_of_moved_cards == board_two_steps_back->get_column(source_column_index).get_number_of_face_up_cards())
            {
                /*
                 * The previous move took all face-up cards from the source column.
                 * Therefore, it is not suspicious.
                 * (Shenanigans with the king are handled by other filters.)
                 */
                return false;
            }
            else
            {
                /*
                 * The previous move took only a portion of the available run.
                 * Therefore, we want to make sure the top-most card NOT taken
                 * is involved with the current move.  Otherwise, the previous
                 * move has no value.
                 */
                deck::card card_exposed_by_run_move{ previous_board->get_column(source_column_index).get_top_card() };
                if (vector<deck::card>{ card_exposed_by_run_move } == current_move->get_cards())
                {
                    // This move uses the exposed card, meaning the previous run move is allowed.
                    // The makes sure the move uses ONLY the exposed card, not as one of a run.
                    // TODO: Should the only allowed follow-up move be column to foundation?
                    return false;
                }
                else
                {
                    /*
                     * This move does not use the exposed card, so filter the sequence of moves.
                     */
                    return true;
                }
            }
        }
        else
        {
            /*
             * Not interested in any scenario except where the second-most-recent move is moving a run from one column to another.
             */
            return false;
        }
    }
}
